packages/files/file_path.pony

```diff
@@ -54,70 +54,78 @@ class val FilePath
     """
 
   new val create(
-    base: (FilePath | AmbientAuth),
+    base: (AmbientAuth | FileAuth),
     path': String,
     caps': FileCaps val = recover val FileCaps .> all() end)
-    ?
   =>
     """
-    Create a new path. The caller must either provide the root capability or an
-    existing FilePath.
+    Create a new path.
 
-    If the root capability is provided, path' will be relative to the program's
-    working directory. Otherwise, it will be relative to the existing FilePath,
-    and the existing FilePath must be a prefix of the resulting path.
+    Unless absolute, path' will be relative to the program's working directory.
 
-    The resulting FilePath will have capabilities that are the intersection of
-    the supplied capabilities and the capabilities on the parent.
+    Capabilities are as provided, unlike with `FilePath.from(...)`.
     """
     caps.union(caps')
+    path = Path.abs(path')
 
-    path = match base
-      | let b: FilePath =>
-        if not b.caps(FileLookup) then
-          error
-        end
+  new val from(
+    base: FilePath,
+    path': String,
+    caps': FileCaps val = recover val FileCaps .> all() end) ?
+  =>
+    """
+    Create a new path from an existing `FilePath`.
 
-        let tmp_path = Path.join(b.path, path')
-        caps.intersect(b.caps)
+    path' is relative to the existing `FilePath`,
+    and the existing `FilePath` must be a prefix of the resulting path.
 
-        if not tmp_path.at(b.path, 0) then
-          error
-        end
-        tmp_path
-      | let b: AmbientAuth =>
-        Path.abs(path')
-      end
+    The resulting `FilePath` will have capabilities that are the intersection of
+    the supplied capabilities and the capabilities of the existing `FilePath`.
+    """
+    caps.union(caps')
+    if not base.caps(FileLookup) then
+      error
+    end
+
+    let tmp_path = Path.join(base.path, path')
+    caps.intersect(base.caps)
+
+    if not tmp_path.at(base.path, 0) then
+      error
+    end
+    path = tmp_path
 
   new val mkdtemp(
-    base: (FilePath | AmbientAuth),
+    base: (AmbientAuth | FileAuth | FilePath),
     prefix: String = "",
-    caps': FileCaps val = recover val FileCaps .> all() end)
-    ?
+    caps': FileCaps val = recover val FileCaps .> all() end) ?
   =>
     """
     Create a temporary directory and returns a path to it. The directory's name
-    will begin with `prefix`. The caller must either provide the root
-    capability or an existing FilePath.
+    will begin with `prefix`.
 
-    If AmbientAuth is provided, pattern will be relative to the program's
+    If `AmbientAuth` or `FileAuth` are provided, pattern will be relative to the program's
     working directory. Otherwise, it will be relative to the existing
-    FilePath, and the existing FilePath must be a prefix of the resulting path.
+    `FilePath`, and the existing `FilePath` must be a prefix of the resulting path.
```

Comment:
I was confused at the mention of the  word `pattern` here, since it's not defined anywhere else. This comes from the [original implementation of `mkdtemp`](https://github.com/ponylang/ponyc/commit/346b5bb20edbc6b90873ad7d8c33585493f846ab), which took a parameter `pattern` instead of the current `prefix`.

I'd probably reword this and say "If `AmbientAuth` or `FileAuth` are provided, the resulting `Filepath` will be relative ....`

We can either do this here or in a separate PR, I don't mind.

Reply:
I will get it fixed now.